src/Install/Requirements.php

```diff
@@ -0,0 +1,129 @@
+ *
+ * This file is about checks for requirements during install of MongoDB module.
+ */
+
+namespace Drupal\mongodb\Install;
+
+/**
+ * Class Requirements implements hook_requirements().
+ */
+class Requirements {
+  public function __construct() {
+    $this->extension_name = '';
+    $this->requirements = array();
+
+    // Ensure translations don't break at install time.
+    $this->t = get_t();
+  }
+
+/**
+ * Common entry point for all requirement checks.
+ */
+  public static function check($phase) {
+    // Check extension version
+    $this->checkExtension();
+
+    // Check server version
+    $this->checkServer($phase);
+
+    return $this->requirements;
+  }
+
+/**
+ * Checks mongodb extension version.
+ */
+  public function checkExtension() {
+    if (extension_loaded('mongodb')) {
+      $this->extension_name = 'mongodb';
+      if (class_exists(MongoClient::class)) {
+        $this->requirements['mongodb_extension'] += array(
+          'value' => t('Properly replaced'),
+          'description' => $t('The mongo-php-adapter library and the ext-mongodb extension are present, replacing the ext-mongo extension.'),
+          'severity' => REQUIREMENT_OK,
+        );
+      }
+      else {
+        $this->requirements['mongodb_extension'] += array(
+          'value' => t('Insufficient'),
+          'description' => $t('The ext-mongodb extension is present, but the mongo-php-adapter is missing, so it cannot replace the old ext-mongo extension.'),
+          'severity' => REQUIREMENT_ERROR,
+        );
+      }
+    }
+    elseif (!extension_loaded('mongo')) {
+      if (class_exists(MongoClient::class)) {
+        $this->requirements['mongodb_extension'] += array(
+          'value' => t('Insufficient'),
+          'description' => $t('The mongo-php-adapter library is present, but the ext-mongodb is missing, so it cannot replace the old ext-mongo extension.'),
+          'severity' => REQUIREMENT_ERROR,
+        );
+      }
+      else {
+        $this->requirements['mongodb_extension'] += array(
+          'value' => t('Insufficient'),
+          'description' => $t('Neither the old ext-mongo nor the new ext-mongodb extension nor the mongo-php-adapter library are present.'),
+          'severity' => REQUIREMENT_ERROR,
+        );
+      }
+    }
+    else {
+      $this->extension_name = 'mongo';
+      if (class_exists(\Alcaeus\MongoDbAdapter\TypeInterface::class)) {
+        $this->requirements['mongodb_library_redundant'] = array(
+          'title' => 'mongo-php-adapter',
+          'value' => t('Redundant'),
+          'description' => t('The mongo-php-adapter library is installed while the old ext-mongo extension is present. This does not hurt, but is somewhat redundant.'),
+          'severity' => REQUIREMENT_WARNING,
+        );
+      }
+    }
+  }
+
+/**
+ * Checks mongodb server version.
+ */
+  public function checkServer($phase) {
+    $this->extension = new ReflectionExtension($this->extension_name);
```

Comment:
$this->extension doesn't seem to be reused outside this method, so it's better (encapsulation) to just use a local  variable.

Also, reflection may fail, so it is necessary to handle the exceptions.